Core/include/Acts/TrackFitting/Chi2Fitter.hpp

```diff
@@ -0,0 +1,734 @@
+            trackStateProxy.effectiveProjector() * residualsFull;  // 2x1
+        // TODO: use detail::calculateResiduals?
+
+        const auto& deriv1 =
+            -2 * Hi.transpose() * covarianceInverse * residuals;  // 6x1
+        const auto& deriv2 =
+            2 * Hi.transpose() * covarianceInverse * Hi;  // 6x6
+
+        result.d1_deriv_sum_collector += deriv1;
+        result.d2_deriv_sum_collector += deriv2;
+
+        result.m_measurements_collector.push_back(parameters(0));
+        result.mcov_measurement_covariance_collector.push_back(
+            covariance(0, 0));
+        result.r_residual_collector.push_back(residuals(0));
+
+        // KF: filtered = predicted + K * (calibrated - H * predicted);
+        trackStateProxy.filtered() =
+            trackStateProxy.predicted();  // in KF, this is set in
+                                          // GainMatrixUpdater.hpp:102, called
+                                          // from filter():
+        trackStateProxy.filteredCovariance() =
+            trackStateProxy.predictedCovariance();
+
+        if (parameters.rows() == 2) {
+          // strip
+          result.m_measurements_collector.push_back(parameters(1));
+          result.r_residual_collector.push_back(residuals(1));
+          result.mcov_measurement_covariance_collector.push_back(
+              covariance(1, 1));
+        }
+
+        //====================================
+
+        // Get and set the type flags
+        auto& typeFlags = trackStateProxy.typeFlags();
+        typeFlags.set(TrackStateFlag::ParameterFlag);
+        if (surface->surfaceMaterial() != nullptr) {
+          typeFlags.set(TrackStateFlag::MaterialFlag);
+        }
+
+        if (not extensions.outlierFinder(trackStateProxy)) {
+          typeFlags.set(TrackStateFlag::MeasurementFlag);
+          ++result.measurementStates;
+        } else {
+          ACTS_INFO("chi2 | Measurement is determined to be an outlier.");
+          typeFlags.set(TrackStateFlag::OutlierFlag);
+          trackStateProxy.data().ifiltered = trackStateProxy.data().ipredicted;
+        }
+
+        // Update state and stepper with post material effects
+        // materialInteractor(surface, state, stepper, postUpdate); // TODO
+        // We count the processed state
+        ++result.processedStates;
+        // Update the number of holes count only when encoutering a
+        // measurement
+        result.measurementHoles = result.missedActiveSurfaces.size();  // TODO
+        // Since we encountered a measurment update the lastMeasurementIndex to
+        // the lastTrackIndex.
+        result.lastMeasurementIndex = result.lastTrackIndex;
+
+      } else if (surface->associatedDetectorElement() != nullptr ||
+                 surface->surfaceMaterial() != nullptr) {
+        // We only create track states here if there is already measurement
+        // detected or if the surface has material (no holes before the first
+        // measurement
```

Comment:
We need to consider the material effects even if there is no measurement on the surface.